db/revision_cache_lru.go

```diff
@@ -272,42 +329,212 @@ func (rc *LRURevisionCache) getValue(docID, revID string, create bool) (value *r
+		return nil
+	}
+
+	key := IDandCV{DocID: docID, Source: cv.SourceID, Version: cv.VersionCAS}
+	rc.lock.Lock()
+	if elem := rc.hlvCache[key]; elem != nil {
+		rc.lruList.MoveToFront(elem)
+		value = elem.Value.(*revCacheValue)
+	} else if create {
+		value = &revCacheValue{id: docID, cv: *cv}
+		newElem := rc.lruList.PushFront(value)
+		rc.hlvCache[key] = newElem
+		for rc.lruList.Len() > int(rc.capacity) {
+			rc.purgeOldest_()
+		}
+	}
+	rc.lock.Unlock()
+	return
+}
+
+// addToRevMapPostLoad will generate and entry in the Rev lookup map for a new document entering the cache
+func (rc *LRURevisionCache) addToRevMapPostLoad(docID, revID string, cv *CurrentVersionVector) {
+	legacyKey := IDAndRev{DocID: docID, RevID: revID}
+	key := IDandCV{DocID: docID, Source: cv.SourceID, Version: cv.VersionCAS}
+
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	// check for existing value in rev cache map (due to concurrent fetch by rev ID)
+	cvElem, cvFound := rc.hlvCache[key]
+	revElem, revFound := rc.cache[legacyKey]
+	if !cvFound {
+		// its possible the element has been evicted if we don't find the element above (high churn on rev cache)
+		// need to return doc revision to caller still but no need repopulate the cache
+		return
+	}
+	// Check if the element in the other lookup cache points to a value that has been updated in both lookup maps already
+	// This protects the following scenario:
+	//	- Process A is lookup via CV and Process B is lookup via revid, assume both process concurrent and both miss cache thus load from bucket
+	//	- Assume for this example both processes manage to 1. update their lookup map, 2. create a new value for the cache, 3. load from bucket to populate this value
+	//	- Now both processes need to update the other lookup map to their respective values
+	//	- Say process A acquires lock on rev cache for this function first, it looks up the key in the rev lookup map and finds it
+	//	assumes duplicate and removes the entry from the map and its respective element from the cache list
+	//	 - Process A continues on and updates the rev map now to point to its own value, process A then release the lock
+	//	 - Process B has been waiting on the lock, now its free it acquires the rev cache lock
+	//	 - Process B looks up key in cv lookup map, find and entry, assumes duplicate and removes the entry from the cv lookup map,
+	//	 and removes the element from teh cache list
+	//	 - We now find ourselves in a situation where both processes have removed both respective cache loads from the cache itself
+	//	 thus no entry in cache for this doc revision/version anymore. This needs to be avoided.
+	if revFound {
+		if cvElem == revElem {
+			// already match, return
+			return
+		}
+		rc.cache[legacyKey] = cvElem
+		rc.lruList.Remove(revElem)
+	}
+
+	// grab elem again using the hlv key in case the element in question has been evicted, in this case we don't want to update cache
+	if elem := rc.hlvCache[key]; elem != nil {
```

Comment:
I don't think this block (from line 402 to 414) is necessary after making the above changes, is it?

Reply:
I think we do, If we get to the point in the code on line 393 and we revFound is false we still need to add it to the rev lookup map, as the item hasn't been found in that map. Upon reading with fresher eyes this morning I realise I can refactor to be part of the if statement to make it better though. I did sanity check my thinking by removing the code altogether and running test but a lot of the test fail without it for example on a test where i'm putting a doc with cv defined in the cache, then lookup by revid it fails to find it.

Reply:
Extra context: Essentially gets exposed when doing a PUT on rev cache. As I have written the code to always insert using the CV (the getValueByCV code path) then subsequently call addToRevMapPostLoad to update the rev lookup. When it gets to this code on line 393 as the code has so far only added to the cv lookup map it doesn't have an entry in the rev lookup. So this code here is needed to add the reference to rev lookup.

Reply:
On Friday we discussed about removing the rev id assignment here. I have no refactored to check if we have the rev id defined on the cvElem and if not we assign rev id. This is because the rev id assignment here is needed for the PUT code path. So to ensure we only acquire the lock when we need it I have refactored to check if the revid is empty on the elem in the list, if so we define it else we leave it just adding to the rev lookup map.